# Building Production Ready RAG Pipeline

In this notebook you will learn to build a Production ready RAG Pipeline on `Attention is All You Need` paper. We will use `Sentence Window Index` to build a basic RAG pipeline and iterate over different parameters to make it production ready.

Following are the steps involved:

1. Download Data
2. Load Data
3. Build Evaluation Dataset.
4. Download `RagEvaluatorPack`.
5. Define LLM, Embedding Model.
6. Build RAG with `Sentence Window` approach.
7. Evaluate RAG Pipeline.
8. Create functions to build index, evaluate.
9. Tune different parameters to improve metrics and make it production ready.

## Setup

Install the libraries.

In [ ]:
!pip install llama-index pypdf torch sentence-transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.7/15.7 MB 52.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 23.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 8.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.0/143.0 kB 11.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.9/75.9 kB 4.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 225.4/225.4 kB 10.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 43.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 14.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.9/76.9 kB 4.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 2.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.4/49.4 kB 3.4 MB/s eta 0:00:00
  Created wheel for sentence-transformers: filename=sentence_transforme

## Download `Attention is all you need` paper.

In [ ]:
!mkdir './data'
!wget --user-agent="Mozilla" "https://arxiv.org/pdf/1706.03762.pdf" -O "./data/attention_is_all_you_need.pdf"

--2023-12-25 10:28:54--  https://arxiv.org/pdf/1706.03762.pdf
Resolving arxiv.org (arxiv.org)... 151.101.195.42, 151.101.131.42, 151.101.67.42, ...
Connecting to arxiv.org (arxiv.org)|151.101.195.42|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2215244 (2.1M) [application/pdf]
Saving to: ‘./data/attention_is_all_you_need.pdf’

./data/attention_is 100%[===================>]   2.11M  --.-KB/s    in 0.07s   

2023-12-25 10:28:54 (29.9 MB/s) - ‘./data/attention_is_all_you_need.pdf’ saved [2215244/2215244]



## Set `OpenAI` keys.

In [ ]:
import nest_asyncio

nest_asyncio.apply()

import os
os.environ['OPENAI_API_KEY'] = 'YOUR OPENAI API KEY'

## Load Data.

We will use first 10 pages and skip paper references in the paper.

In [ ]:
from llama_index import SimpleDirectoryReader

data = SimpleDirectoryReader('./data/').load_data()

documents = data[:10]

## Generate Evaluation dataset using `RagDatasetGenerator` and `GPT-4`

In [ ]:
from llama_index import ServiceContext
from llama_index.llms import OpenAI
from llama_index.llama_dataset.generator import RagDatasetGenerator

gpt4 = OpenAI(model='gpt-4', temperature=0.1)
service_context_gpt4 = ServiceContext.from_defaults(llm=gpt4)

dataset_generator = RagDatasetGenerator.from_documents(
    documents,
    service_context=service_context_gpt4,
    num_questions_per_chunk=2,
    show_progress=True,
)

eval_dataset = dataset_generator.generate_dataset_from_nodes()

Parsing nodes:   0%|          | 0/10 [00:00<?, ?it/s]

100%|██████████| 2/2 [00:04<00:00,  2.45s/it]


## Download `RagEvaluatorPack` for evaluation.

In [ ]:
from llama_index.llama_pack import download_llama_pack

RagEvaluatorPack = download_llama_pack(
  "RagEvaluatorPack", "./rag_evaluator_pack"
)

## Define LLM.

In [ ]:
from llama_index.llms import OpenAI

llm = OpenAI(model="gpt-3.5-turbo", temperature=0.1)

## Define Embedding Model

In [ ]:
embed_model = "local:BAAI/bge-small-en-v1.5"

## Build RAG pipeline with `SentenceWindow`

In [ ]:
from llama_index.node_parser import SentenceWindowNodeParser

# create the sentence window node parser w/ default settings
node_parser = SentenceWindowNodeParser.from_defaults(
    window_size=1,
    window_metadata_key="window",
    original_text_metadata_key="original_text",
)

In [ ]:
from llama_index import ServiceContext

sentence_context = ServiceContext.from_defaults(
    llm=llm,
    embed_model=embed_model,
    node_parser=node_parser,
)

config.json:   0%|          | 0.00/743 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/133M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/366 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/711k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

In [ ]:
from llama_index import Document

document = Document(text="\n\n".join([doc.text for doc in documents]))

In [ ]:
from llama_index import VectorStoreIndex

sentence_index = VectorStoreIndex.from_documents(
    [document], service_context=sentence_context
)

In [ ]:
from llama_index.indices.postprocessor import SentenceTransformerRerank

rerank = SentenceTransformerRerank(
    top_n=2, model="BAAI/bge-reranker-base"
)

config.json:   0%|          | 0.00/799 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.11G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/443 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/279 [00:00<?, ?B/s]

In [ ]:
from llama_index.indices.postprocessor import MetadataReplacementPostProcessor

postproc = MetadataReplacementPostProcessor(
    target_metadata_key="window"
)

In [ ]:
query_engine = sentence_index.as_query_engine(
    similarity_top_k=2, node_postprocessors=[postproc, rerank]
)

In [ ]:
response = query_engine.query('is the paper from google research?')
print(response)

Yes.


## Evaluate RAG pipeline

In [ ]:
rag_evaluator_pack = RagEvaluatorPack(
    rag_dataset=eval_dataset,
    query_engine=query_engine
)

base_benchmark = await rag_evaluator_pack.arun(
    batch_size=10,  # batches the number of openai api calls to make
    sleep_time_in_seconds=1,  # seconds to sleep before making an api call
)


Batch processing of predictions: 100%|██████████| 10/10 [00:15<00:00,  1.59s/it]
Batch processing of evaluations:  87%|████████▋ | 10/11.5 [01:23<00:12,  8.34s/it]


In [ ]:
base_benchmark

rag                            base_rag
metrics                                
mean_correctness_score         3.300000
mean_relevancy_score           0.800000
mean_faithfulness_score        1.000000
mean_context_similarity_score  0.910959

## Create Functions to build RAG pipeline and Evaluation.

This will make the process of iterating easier for evaluation.

In [ ]:
def build_index(
    documents,
    llm=OpenAI(model='gpt-3.5-turbo', temperature=0.1),
    embed_model="local:BAAI/bge-small-en-v1.5",
    sentence_window_size=3,
):
    # create the sentence window node parser w/ default settings
    node_parser = SentenceWindowNodeParser.from_defaults(
        window_size=sentence_window_size,
        window_metadata_key="window",
        original_text_metadata_key="original_text",
    )
    sentence_context = ServiceContext.from_defaults(
        llm=llm,
        embed_model=embed_model,
        node_parser=node_parser,
    )
    sentence_index = VectorStoreIndex.from_documents(
        documents, service_context=sentence_context
    )

    return sentence_index


def setup_query_engine(
    sentence_index, similarity_top_k=2, rerank_top_n=2, is_rerank = False
):
    # define postprocessors
    postproc = MetadataReplacementPostProcessor(target_metadata_key="window")
    if is_rerank:
      rerank = SentenceTransformerRerank(
          top_n=rerank_top_n, model="BAAI/bge-reranker-base"
      )
      query_engine = sentence_index.as_query_engine(
          similarity_top_k=similarity_top_k, node_postprocessors=[postproc, rerank]
      )
    else:
      query_engine = sentence_index.as_query_engine(
          similarity_top_k=similarity_top_k, node_postprocessors=[postproc]
      )
    return query_engine

async def evaluate(query_engine):
  rag_evaluator_pack = RagEvaluatorPack(
      rag_dataset=eval_dataset,
      query_engine=query_engine
  )

  benchmark_df = await rag_evaluator_pack.arun(
      batch_size=10,  # batches the number of openai api calls to make
      sleep_time_in_seconds=1,  # seconds to sleep before making an api call
  )

  return benchmark_df

def build_index_query_engine(window_size,
                             similarity_top_k,
                             rerank_similarity_top_k,
                             rerank_top_k):
  sentence_index = build_index(
      [document],
      sentence_window_size=window_size,
  )

  query_engine = setup_query_engine(sentence_index,
                                      similarity_top_k=similarity_top_k)

  query_engine_rerank = setup_query_engine(sentence_index,
                                            similarity_top_k=rerank_similarity_top_k,
                                            rerank_top_n=rerank_top_k,
                                            is_rerank=True)
  return sentence_index, query_engine, query_engine_rerank

In [ ]:
index, query_engine, query_engine_rerank = build_index_query_engine(1, 2, 4, 2)

In [ ]:
base_benchmark = await evaluate(query_engine)

Batch processing of predictions: 100%|██████████| 10/10 [00:04<00:00,  2.09it/s]
Batch processing of evaluations:  87%|████████▋ | 10/11.5 [01:10<00:10,  7.00s/it]


In [ ]:
base_benchmark_rerank = await evaluate(query_engine_rerank)

Batch processing of predictions: 100%|██████████| 10/10 [00:28<00:00,  2.86s/it]
Batch processing of evaluations:  87%|████████▋ | 10/11.5 [01:10<00:10,  7.05s/it]


In [ ]:
base_benchmark

rag                            base_rag
metrics                                
mean_correctness_score         3.275000
mean_relevancy_score           0.800000
mean_faithfulness_score        1.000000
mean_context_similarity_score  0.907221

In [ ]:
base_benchmark_rerank

rag                            base_rag
metrics                                
mean_correctness_score          3.15000
mean_relevancy_score            0.85000
mean_faithfulness_score         1.00000
mean_context_similarity_score   0.91542

From the metrics we can observe that `correctness` metric is lower (maximum 5) and including `reranker` improved metrics though it decreased the `correctness` metric.

Interesting to see there are no hallucinations as `faithfulness` metric is 1.0

## Tune parameters to make it production ready.

Let's aim to get `correctness` score of `4.5` and `relevancy` score of more than `0.9`.

### Experiment 1:

Let's increase window size and see if we can improve correctness as it gives more surrounding context.

In [ ]:
index, query_engine, query_engine_rerank = build_index_query_engine(3, 2, 4, 2)

In [ ]:
benchmark = await evaluate(query_engine)

Batch processing of predictions: 100%|██████████| 10/10 [00:04<00:00,  2.23it/s]
Batch processing of evaluations:  87%|████████▋ | 10/11.5 [01:29<00:13,  8.96s/it]


In [ ]:
benchmark_rerank = await evaluate(query_engine_rerank)

Batch processing of predictions: 100%|██████████| 10/10 [00:54<00:00,  5.50s/it]
Batch processing of evaluations:  87%|████████▋ | 10/11.5 [01:28<00:13,  8.88s/it]


In [ ]:
benchmark

rag                            base_rag
metrics                                
mean_correctness_score         3.775000
mean_relevancy_score           0.950000
mean_faithfulness_score        1.000000
mean_context_similarity_score  0.917831

In [ ]:
benchmark_rerank

rag                            base_rag
metrics                                
mean_correctness_score         3.875000
mean_relevancy_score           0.950000
mean_faithfulness_score        1.000000
mean_context_similarity_score  0.928741

The metrics did improve.

### Experiment 2:

Let'r increase `similarity_top_k` and `reranker_top_n` values and see if getting more relevant contexts improves the result.

In [ ]:
index, query_engine, query_engine_rerank = build_index_query_engine(3, 4, 8, 4)

In [ ]:
benchmark = await evaluate(query_engine)

Batch processing of predictions: 100%|██████████| 10/10 [00:05<00:00,  1.80it/s]
Batch processing of evaluations:  87%|████████▋ | 10/11.5 [01:33<00:13,  9.31s/it]


In [ ]:
benchmark_rerank = await evaluate(query_engine_rerank)

Batch processing of predictions: 100%|██████████| 10/10 [02:17<00:00, 13.78s/it]
Batch processing of evaluations:  87%|████████▋ | 10/11.5 [01:21<00:12,  8.10s/it]


In [ ]:
benchmark

rag                            base_rag
metrics                                
mean_correctness_score         4.100000
mean_relevancy_score           0.950000
mean_faithfulness_score        1.000000
mean_context_similarity_score  0.924449

In [ ]:
benchmark_rerank

rag                            base_rag
metrics                                
mean_correctness_score         4.275000
mean_relevancy_score           0.900000
mean_faithfulness_score        1.000000
mean_context_similarity_score  0.934464

We are close to our goal. We reached a `correctness` score of `4.275` and `relevancy` score of 0.9. The `context similarity score` also improved

### Experiment 3:

Let's now increase window size and see if it improves the metrics.

In [ ]:
index, query_engine, query_engine_rerank = build_index_query_engine(5, 4, 8, 4)

In [ ]:
benchmark = await evaluate(query_engine)

Batch processing of predictions: 100%|██████████| 10/10 [00:07<00:00,  1.36it/s]
Batch processing of evaluations:  87%|████████▋ | 10/11.5 [01:16<00:11,  7.67s/it]


In [ ]:
benchmark_rerank = await evaluate(query_engine_rerank)

Batch processing of predictions: 100%|██████████| 10/10 [03:14<00:00, 19.49s/it]
Batch processing of evaluations:  87%|████████▋ | 10/11.5 [01:22<00:12,  8.26s/it]


In [ ]:
benchmark

rag                            base_rag
metrics                                
mean_correctness_score         4.250000
mean_relevancy_score           1.000000
mean_faithfulness_score        1.000000
mean_context_similarity_score  0.926042

In [ ]:
benchmark_rerank

rag                            base_rag
metrics                                
mean_correctness_score         4.250000
mean_relevancy_score           1.000000
mean_faithfulness_score        1.000000
mean_context_similarity_score  0.933203

We have reached a `relevency` score of `1.0`.

### Experiment 4:

Let'r increase `similarity_top_k` and `reranker_top_n` values and see if getting more relevant contexts improves the result.

In [ ]:
index, query_engine, query_engine_rerank = build_index_query_engine(5, 6, 12, 6)

In [ ]:
benchmark_rerank = await evaluate(query_engine_rerank)

Batch processing of predictions: 100%|██████████| 10/10 [05:02<00:00, 30.20s/it]
Batch processing of evaluations:  87%|████████▋ | 10/11.5 [01:20<00:12,  8.09s/it]


In [ ]:
benchmark_rerank

rag                            base_rag
metrics                                
mean_correctness_score          4.50000
mean_relevancy_score            1.00000
mean_faithfulness_score         1.00000
mean_context_similarity_score   0.93266

We have reached our goal of `correctness` score of `4.5` and `relevancy` score of `1.0` (>0.9).

## Observation:

In this project, we looked into building RAG Pipeline, evaluation dataset and tuning different parameters to make it production ready. It should be observed that `reranker` improved metrics in most of the experiments.

Please do remember that we have various other metrics like `chunk_size`, `chunk_overlap`, `embedding model`, `LLM` to experiment.